### ISSUE TITLE
Barbarians no longer attack

### ISSUE THREAD

**Issue by DeStup** [yairm210/Unciv#10009]
Hello!

OC:
Android, Windows_64

Version:
After patch 4.7.18, vanilla

Description:
Barbarians no longer attack my units.

Possible reason from the patch notes:
"By tuvus:
AI nuke improvement
AI move units closer to enemy first in wartime
Improved AI attack targetting"

1. Start any game on the patch after 4.7.18
2. Approach the barbarians

![Unciv_TviAmxWeHa](https://github.com/yairm210/Unciv/assets/59770002/78173f77-a1f0-4165-a412-b4ae00f786aa)

**Comment by tuvus** [comment]
I have also noticed this, but I didn't fully realize that it was the case all the time. I thought I read somewhere that the barbarians use a different piece of code to attack. I'll look into it. Thanks for posting this issue!

**Comment by tuvus** [comment]
Through testing, I know see that barbarians did indeed attack but only very very rarely. My changes to the AttackAI made the AI a lot safer. Previously barbarians attack a target in range, regardless of how good it is. Since they are very often underpowered they end up finding that they shouldn't attack any target.

This does raise the concern that underpowered Civs won't fight civs that are much stronger.

Also, are melee barbarians supposed to walk aimlessly around with low health never attacking?

**Comment by yairm210** [comment]
Barbarians are supposed to heal before wandering, for sure. 
And units should attack even if they're not as strong, otherwise one strong unit will annihilate an army of weaker ones

**Comment by tuvus** [comment]
Ooooh, I get it now! The biggest problem had nothing to do with the value system that I implemented! While implementing the new attackValueing method, I used a for loop to loop through each possibility and pick the best along with its value. Then, I check to see if the value is bigger than a minimum to see if it should attack at all. In review, it was recommended that I use .maxBy {} as this seems to be exactly the same. Except .maxBy {} shortcuts if there is only one element, meaning that it won't update the highest attack value and, therefore won't attack at all.
This behavior made armies fight and solo combatants peacefully ignore eachother.
